crmprtd/bc_hydro/download.py

```diff
@@ -0,0 +1,116 @@
+''' Downloads data from BC hyrdo
+
+BC Hydro posts a rolling window (3 months) observing hourly data
+once a week. Data is in txt files.
+
+It is recommended to run this script once every 3 months to collect
+all available data, supplemented with aggressive monitoring and
+alerting for errors. If the script is run less than 3 months you
+will miss data.
+'''
+
+import pysftp
+from tempfile import TemporaryDirectory
+import logging
+import os
+import sys
+import re
+from zipfile import ZipFile
+from argparse import ArgumentParser
+from datetime import date, timedelta
+
+from crmprtd import logging_args, setup_logging, common_auth_arguments
+
+log = logging.getLogger(__name__)
+
+
+def download(username, gpg_private_key, ftp_server, ftp_dir, start_date,
+             end_date):
+
+    # Connect FTP server and retrieve file
+    try:
+        sftp = pysftp.Connection(ftp_server, username=username,
+                                 private_key=gpg_private_key)
+    except Exception:
+        log.exception("Invalid ftp authentication")
+        sys.exit(1)
+
+    try:
+        with TemporaryDirectory() as tmp_dir:
+            os.chdir(os.path.expanduser(tmp_dir))
+            range = DateRange(int(start_date), int(end_date), sftp)
+            sftp.walktree(ftp_dir, range.matchFile, matchDir, unknownType)
+            for filename in os.listdir(os.getcwd()):
+                name, extension = os.path.splitext(filename)
+                if extension == '.zip':
+                    zip_file = ZipFile(filename, 'r')
+                    for name in zip_file.namelist():
+                        txt_file = zip_file.open(name)
+                        sys.stdout.buffer.write(txt_file.read())
+
+    except IOError:
+        log.exception("Unable to download or open files")
+        sys.exit(1)
+
+
+class DateRange():
```

Comment:
Not meaning to be snarky, but whenever I see a class definition with two methods, and one of them is `__init__` I can't help but refer back to this [classic PyCon talk](https://www.youtube.com/watch?v=o9pEzgHorH0) which advises against such a pattern (see *at least* 2:20-3:20).

You can probably do the same thing with a single method that has four args:

```python
def matchFile(sftp, filename, start_date, end_date):
    ...
```

and it will be more stateless and significantly more testable.

Reply:
I understand. I did this because I wasn't sure how to pass the filename and 3 more arguments to matchFile through connection.walktree and this was my work around.

Reply:
Oh OK, that makes sense. I *think* what you want is [`functools.partial`](https://docs.python.org/3/library/functools.html#functools.partial). So define your function `matchFile()` with signature:

```python3
def matchFile(filename, start_date, end_date):
    ...
```

And then before the `walktree` call, "freeze" the arguments with the parameters that you have:

```python3
callback = functools.partial(matchFile, start_date, end_date)
sftp.walktree(ftp_dir, matchFile, no_op, no_op)
```

I *think* that's a little more elegant (and will hopefully work), but I'll let you work out the exact implementation.